### ISSUE TITLE
Project prototype

### ISSUE THREAD

**Issue by visortelle** [tealtools/puls#1]
## Why

- It's hard to quickly run and test Apache Pulsar in different configurations on a developer machine. Some cases can't be tested with Pulsar Standalone like geo-replication or tiered storage.
- Making it simple to run Pulsar could make Pulsar more attractive for developers.
- We at Teal Tools develop [UI for Pulsar](https://dekaf.io) and need such a tool to quickly test some features.
- Someone created a [similar project for Apache Kafka](https://www.linkedin.com/posts/kiryl-valkovich_creating-a-docker-composeyml-file-with-kafka-docker-composer-activity-7187331102959681536-fw0q?utm_source=share&utm_medium=member_desktop). I thought about having something similar for Pulsar for a long time.

## Goals

- Provide the simplest possible way to run Apache Pulsar. Ideally, the user should run `pulsar-compose run --num-clusters 2` and in about 2 minutes get a fully functional geo-replicated Pulsar instance up and running.
- Run 3rd party components needed to demonstrate some Pulsar features like tiered storage or IO
- Single cluster should require as low memory as possible by default
- Optionally run UI
- Optionally run Prometheus and Grafana
- If it possible, create an extension for testcontainers. If not, wrap it into a separate library for testing.

### Personal goals

- Try Rust in action with a simple pet project
- Better familiarize myself with running a geo-replicated Pulsar instance from scratch
- Simplify the development of https://dekaf.io/ by providing simple ways to spin up clusters in different configurations

## Non-goals

- Make it secure
- Make it for running Pulsar in production
- Make it very flexible in terms of configuration options

## Roadmap

- [x] Run Pulsar components
  - [x] Broker
  - [x] Bookie
  - [x] ZooKeeper
  - [x] Proxy
- [x] Geo-replication
  - [x] Create multiple clusters
  - [x] Connect clusters as a single Pulsar instance
- [ ] Pulsar Function workers
- [ ] Minio (S3) for testing Tiered storage
- [ ] Some database for Pulsar IO testing. Most likely MongoDB or Postgres
- [ ] Prometheus and Grafana
- [x] UI
- [ ] Ensure that it runs on all platforms
  - [ ] Linux
  - [ ] Windows (WSL2)
  - [x] MacOS
- [ ] Build release artifacts on CI
- [x] Better CLI
  - [x] Produce a `docker-compose.yml` file by running `pulsar-compose generate`.
  - [x] Create instance `pulsar-compose create --num-clusters 2 --num-brokers 5 my-pulsar`
  - [x] `pulsar-compose run` without additional arguments should generate a random instance name, create it, and run it.
  - [x] `pulsar-compose run my-pulsar` should run a created instance.
  - [x] List instances `pulsar-compose ls`. Shows a list of instances and their statuses.
  - [x] Describe instance `pulsar-compose describe my-pulsar`. Show instance configuration and status.
  - [x] Stop running instance `pulsar-compose stop my-pulsar`.
  - [x] Delete instance and its data `pulsar-compose delete my-pulsar`.
  - [x] Config files that CLI can write and read

  ```
  pulsarInstance:
    name: test-pulsar
    numClusters: 2
    numBrokersPerCluster: 2
    numBookiesPerCluster: 3

    minio:
      enabled: true
    postgres:
      enabled: true
  ```
- [ ] Resource consumption optimizations
- [ ] Mention as a option for dev here: https://github.com/apache/pulsar/issues/8671
- [x] Shorter command name? `pulsar-compose` is quite long.

**Comment by visortelle** [comment]
A single instance with 3 zookeepers, 3 brokers, 3 bookies, and a pulsar-proxy requires about 2CPUs and 3GB RAM. The CPU consumption is quite high even when idle.

On my old MacBook Pro 2018 with 16GB RAM, I was able to run two Pulsar instances of such size + apache/pulsar repo in IntelliJ Idea + Chrome + multiple apps in the background like Slack, without any visible discomfort. Running an additional third instance makes the system freezes are very noticeable.

I wonder about which default instance size be better:
- 3 replicas for each component are resource-heavy. On one side, most developers expect to have more than 16GB RAM in 2024. Conversely, beginner developers who are also an important audience for Pulsar popularization may not afford pricey laptops.
- 1 replica for each component by default makes it not much different from Pulsar Standalone unless you want to run multiple clusters. This way it would require passing additional arguments to make `pulsar-compose` benefits visible.
- As a compromise, 2 replicas for each component, 1 replica for zookeeper by default. This way I'll be able to run 3 Pulsar clusters on my old MBP 😀 In the same way I'll be able to see how my app behaves in an environment closer to production than with Pulsar Standalone.

Related:
- https://github.com/apache/pulsar/issues/14015
- https://github.com/apache/pulsar/issues/4722

**Comment by visortelle** [comment]
Rust is a great scripting language 😄

- Error messages are very helpful. When I try to use some feature from another language, the compiler recognizes it and tells me that this feature doesn't exist in Rust or provides me an alternative option on how to do what I want.
- Tooling is brilliant. Intuitive, fast, I got zero WTFs per hour.
- No Java-like exceptions.
- All packages I encountered have very good documentation. Probably it's somehow related to the tooling quality, but I didn't see such documentation quality anywhere else.
- Just use `.clone()` everywhere and you'll be very productive. ok for "scripts" or CLI.

---

But there are some flaws too:
- What I miss is a true cross-compilation and something like [goreleaser](https://goreleaser.com/).
- Mixed feelings about macro.
- Even don't try to use any FP tricks. The only thing you'll get is a headache.
